src/main/java/org/jenkinsci/plugins/workflow/steps/durable_task/DurableTaskStep.java

```diff
@@ -347,17 +401,67 @@ private void check() {
             }
         }
 
+        // called remotely from HandlerImpl
+        @Override public void exited(int exitCode, byte[] output) throws Exception {
+            try {
+                getContext().get(TaskListener.class);
+            } catch (IOException | InterruptedException x) {
+                LOGGER.log(Level.FINE, "asynchronous exit notification with code " + exitCode + " in " + remote + " on " + node + " ignored since step already seems dead", x);
+                return;
+            }
+            LOGGER.log(Level.FINE, "asynchronous exit notification with code {0} in {1} on {2}", new Object[] {exitCode, remote, node});
+            if (returnStdout && output == null) {
+                getContext().onFailure(new IllegalStateException("expected output but got none"));
+                return;
+            } else if (!returnStdout && output != null) {
+                getContext().onFailure(new IllegalStateException("did not expect output but got some"));
+                return;
+            }
+            recurrencePeriod = 0;
+            if (returnStatus || exitCode == 0) {
+                getContext().onSuccess(returnStatus ? exitCode : returnStdout ? new String(output, encoding) : null);
+            } else {
+                if (returnStdout) {
+                    listener().getLogger().write(output); // diagnostic
+                }
+                getContext().onFailure(new AbortException("script returned exit code " + exitCode));
+            }
+        }
+
         @Override public void onResume() {
-            setupTimer();
+            ws = null; // find it from scratch please, rewatching as needed
+            setupTimer(MIN_RECURRENCE_PERIOD);
         }
 
-        private void setupTimer() {
-            recurrencePeriod = MIN_RECURRENCE_PERIOD;
+        private void setupTimer(long initialRecurrencePeriod) {
+            recurrencePeriod = initialRecurrencePeriod;
             task = THREAD_POOL.schedule(this, recurrencePeriod, TimeUnit.MILLISECONDS);
         }
 
         private static final long serialVersionUID = 1L;
 
     }
 
+    private static class HandlerImpl extends Handler {
+
+        private static final long serialVersionUID = 1L;
+
+        private final ExecutionRemotable execution;
+        private final TaskListener listener;
+
+        HandlerImpl(Execution execution, FilePath workspace, TaskListener listener) {
+            this.execution = workspace.getChannel().export(ExecutionRemotable.class, execution);
+            this.listener = listener;
+        }
+
+        @Override public void output(InputStream stream) throws Exception {
+            IOUtils.copy(stream, listener.getLogger());
```

Comment:
> other network errors: coverage unknown

😒

Reply:
This implementation means it's *absolutely* a requirement that the invoker know to re-send log content if it fails as the result of transitory network conditions.

Reply:
I do not think so. Network conditions are unlikely to be “transitory”. If there are some lost packets or whatever, TCP will recover and the application layer is not going to know any different.

If the socket is broken—or there is otherwise a serious error in the Remoting transport, say Kafka (under development)—we hope the `Channel` will get closed somehow, and we may subsequently reconnect and resume log transmission. In such cases, we are making a best effort to resume the log where we left off, but we cannot know exactly where that was—the source channel position is an estimate. At worst we lose, or duplicate, a bit of the log file, which is hardly a serious matter when a build is a low-value object (and there is a good chance this build would wind up failing anyway).

At any rate such considerations really only apply to on-master file storage; cloud storage will have completely different characteristics.

Reply:
> At any rate such considerations really only apply to on-master file storage; cloud storage will have completely different characteristics.

Actually, it's even worse in Cloud environments. 

I'll let Google make my argument here about how to design resilient distributed systems and exactly *how little* you can trust the network: http://www.hpcs.cs.tsukuba.ac.jp/~tatebe/lecture/h23/dsys/dsd-tutorial.html#Design

> Both clients and servers must be able to deal with unresponsive senders/receivers.

> When you design distributed systems, you have to say, "Failure happens all the time." So when you design, you design for failure. It is your number one concern. What does designing for failure mean? One classic problem is partial failure. If I send a message to you and then a network failure occurs, there are two possible outcomes. One is that the message got to you, and then the network broke, and I just didn't get the response. The other is the message never got to you because the network broke before it arrived.

I'll out of band cite a couple issues for you too.

Reply:
> it's even worse in Cloud environments

It could be better or much worse, depending on the “cloud”, but at any rate it is irrelevant for purposes of this PR since the behavior is up the provider of the `TaskListener`.

Reply:
Hmm, this is actually worse than it seems, because apparently the `PrintStream` returned by `getLogger` doesn't even throw an exception on failed write operations performed by `IOUtils.copy(stream, listener.getLogger());` - it just sets a 'trouble' flag you have to fetch via 'checkError' :-/

Reply:
Thinking about this a bit deeper, I really think this is broken (a hard 🐛 ) but primarily because of the PrintStream having all-but nonexistent error handling.  There *must* be some way to signal and propagate errors from this process so implementations can signal "hey this thing is broken" and it can get bubbled up. 

I would have deep, deep reservations about treating the log as without value -- IMO half the value of Cloud storage for logs is that they can be audited, and if we don't make *some* effort to deal with failures, we've just lot all ability to audit.

Reply:
> the value of Cloud storage for logs is that they can be audited, and if we don't make some effort to deal with failures

My only concern right now is tolerable behavior for the built-in Remoting (to master file) log sink in the face of corner cases. It is OK to lose a bit of text when the `Channel` is dying or dead: we likely already were in the polling implementation, and the build is often going to fail anyway with an error of no interest to developers. It is not OK to have the build hang indefinitely, a master `Thread` consumed, etc.—things which happened in earlier iterations of the polling implementation and are now I think defended against in most cases.

Cloudy storage of logs is a separate case. Quality of service levels are up to that implementation, not this generic code. I would envision that a production implementation would make some effort to buffer and resend messages in the background, giving up at some point. If it has some evidence that an outage is somehow tied to the state of the agent JVM and that an agent restart would actually solve something, then it could override `PrintStream.write` to throw `RuntimeException`, though I doubt this would be productive.

More of interest is a routine `ChannelClosedException` during master restart, for which we would prefer to have exact log content if at all possible. Not sure if that is effectively testable but can try it. One trick which may help is to extract the `FilterOutputStream.out`.

Reply:
> More of interest

(resolved)